File: c++/leetcode/501. Find Mode in Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<int> ans;
    int _max = 0, count = 0, pre = 0;
    void preorder(TreeNode*);
public:
    vector<int> findMode(TreeNode* root) {
        preorder(root);
        return ans;
    }
};

void Solution::preorder(TreeNode* root) {
    if(!root) return;
    
    preorder(root -> left);
     
    count = pre == root -> val ? ++count : 1;
    
    if(count > _max) {
        ans.clear();
        ans.push_back(root -> val);
        _max = count;
    } 
    else if(count == _max) {
        ans.push_back(root -> val);
    }
    
    pre = root -> val;
    
    preorder(root -> right);
}

File: javascript/leetcode/501. Find Mode in Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findMode = function (root) {
  let ans = [];
  let count = 0,
    _max = 0,
    prev = 0;

  const preorder = (root) => {
    if (!root) return;

    preorder(root.left);

    count = prev == root.val ? ++count : 1;

    if (count > _max) {
      ans = [root.val];
      _max = count;
    } else if (count === _max) ans.push(root.val);

    prev = root.val;

    preorder(root.right);
  };

  preorder(root);
  return ans;
};
